aesd-char-driver/main.c

```diff
@@ -0,0 +1,269 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/syscalls.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("James Bohn");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    PDEBUG("open");
+
+    // save a pointer
+    filp->private_data = &aesd_device;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+
+    // didn't do anything in open, don't need to do anything in release
+
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t bytes_read = 0;
+    struct aesd_dev *dev = (struct aesd_dev *)filp->private_data;
+    size_t offset = 0;
+    struct aesd_buffer_entry *entry;
+
+    if(buf == NULL || filp == NULL || f_pos == NULL){
```

Comment:
@jbohn3353 nice defensive programming here but I think you can assume `filp` won't be null (and you do at line 53).  See https://github.com/cu-ecen-aeld/ldd3/blob/master/scull/main.c#L339 for expectations the scull driver has on file operations functions.